File: app/src/main/java/com/elegion/android/ui/splash/SplashActivity.java
package com.elegion.android.ui.splash;

import android.os.Bundle;
import android.os.CountDownTimer;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.elegion.android.R;
import com.elegion.android.data.Repository;
import com.elegion.android.ui.base.activity.BaseActivity;
import com.elegion.android.ui.features.FeaturesActivity;
import com.elegion.android.ui.login.LoginActivity;

/**
 * @author mikhail barannikov
 */
public class SplashActivity extends BaseActivity implements SplashView {
    private static final int SPLASH_DURATION = 500;

    @InjectPresenter
    SplashPresenter mPresetner;

    private CountDownTimer mTimer;
    private boolean mTimerFinished;

    @ProvidePresenter
    SplashPresenter providePresenter() {
        return new SplashPresenter(Repository.get(this));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.ac_splash);
    }

    @Override
    protected void onResume() {
        super.onResume();
        cancelTimer();
        mTimer = getSplashTimer();
        mTimer.start();
    }

    @Override
    protected void onPause() {
        cancelTimer();
        super.onPause();
    }

    private void cancelTimer() {
        mTimerFinished = false;
        if (mTimer != null) {
            mTimer.cancel();
            mTimer = null;
        }
    }

    private CountDownTimer getSplashTimer() {
        return new CountDownTimer(SPLASH_DURATION, SPLASH_DURATION) {
            @Override
            public void onTick(long millisUntilFinished) {
                // nothing
            }

            @Override
            public void onFinish() {
                mTimerFinished = true;
                openNextActivity();
            }
        };
    }

    private void openNextActivity() {
        if (mTimerFinished) {
            mPresetner.timerFinish();
        }
    }

    @Override
    public void openLogin() {
        startActivity(LoginActivity.makeIntent(this));
        supportFinishAfterTransition();
    }

    @Override
    public void openFeatures() {
        startActivity(FeaturesActivity.makeIntent(this));
        supportFinishAfterTransition();
    }
}

File: app/src/main/kotlin/com/elegion/android/template/ui/splash/SplashActivity.kt
package com.elegion.android.template.ui.splash

import android.os.CountDownTimer

import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.elegion.android.template.R
import com.elegion.android.template.data.Repository
import com.elegion.android.template.ui.base.activity.BaseActivity
import com.elegion.android.template.ui.onboarding.OnBoardingActivity

class SplashActivity : BaseActivity(), SplashView {

    @InjectPresenter
    internal lateinit var presenter: SplashPresenter

    private var timer: CountDownTimer? = null
    private var timerFinished: Boolean = false

    @ProvidePresenter
    internal fun providePresenter(): SplashPresenter = SplashPresenter(Repository.get(this))

    override fun getLayout(): Int = R.layout.ac_splash

    override fun onResume() {
        super.onResume()
        cancelTimer()
        timer = createSplashTimer().start()
    }

    override fun onPause() {
        cancelTimer()
        super.onPause()
    }

    private fun cancelTimer() {
        timerFinished = false
        if (timer != null) {
            timer!!.cancel()
            timer = null
        }
    }

    private fun createSplashTimer(): CountDownTimer {
        return object : CountDownTimer(SPLASH_DURATION.toLong(), SPLASH_DURATION.toLong()) {
            override fun onTick(millisUntilFinished: Long) {
                // nothing
            }

            override fun onFinish() {
                timerFinished = true
                openNextActivity()
            }
        }
    }

    private fun openNextActivity() {
        if (timerFinished) {
            presenter.timerFinish()
        }
    }

    override fun openOnBoarding() {
        startActivity(OnBoardingActivity.makeIntent(this))
        supportFinishAfterTransition()
    }

    override fun openWallet() {
        supportFinishAfterTransition()
    }

    companion object {
        private const val SPLASH_DURATION = 500
    }
}
